x-pack/plugins/actions/server/builtin_action_types/cases_webhook/service.ts

```diff
@@ -0,0 +1,257 @@
+
+export const createExternalService = (
+  actionId: string,
+  { config, secrets }: ExternalServiceCredentials,
+  logger: Logger,
+  configurationUtilities: ActionsConfigurationUtilities
+): ExternalService => {
+  const {
+    createCommentJson,
+    createCommentMethod,
+    createCommentUrl,
+    createIncidentJson,
+    createIncidentMethod,
+    createIncidentResponseKey,
+    createIncidentUrl: createIncidentUrlConfig,
+    getIncidentResponseCreatedDateKey,
+    getIncidentResponseExternalTitleKey,
+    getIncidentResponseUpdatedDateKey,
+    getIncidentUrl,
+    hasAuth,
+    headers,
+    incidentViewUrl,
+    updateIncidentJson,
+    updateIncidentMethod,
+    updateIncidentUrl,
+  } = config as CasesWebhookPublicConfigurationType;
+  const { password, user } = secrets as CasesWebhookSecretConfigurationType;
+  if (
+    !getIncidentUrl ||
+    !createIncidentUrlConfig ||
+    !incidentViewUrl ||
+    !updateIncidentUrl ||
+    (hasAuth && (!password || !user))
+  ) {
+    throw Error(`[Action]${i18n.NAME}: Wrong configuration.`);
+  }
+
+  const createIncidentUrl = removeSlash(createIncidentUrlConfig);
+
+  const axiosInstance = axios.create({
+    ...(hasAuth && isString(secrets.user) && isString(secrets.password)
+      ? { auth: { username: secrets.user, password: secrets.password } }
+      : {}),
+    headers: {
+      ['content-type']: 'application/json',
+      ...(headers != null ? headers : {}),
+    },
+  });
+
+  const getIncident = async (id: string): Promise<GetIncidentResponse> => {
+    try {
+      const res = await request({
+        axios: axiosInstance,
+        url: renderMustacheStringNoEscape(getIncidentUrl, {
+          external: {
+            system: {
+              id,
+            },
+          },
+        }),
+        logger,
+        configurationUtilities,
+      });
+
+      throwIfResponseIsNotValidSpecial({
+        res,
+        requiredAttributesToBeInTheResponse: [
+          getIncidentResponseCreatedDateKey,
+          getIncidentResponseExternalTitleKey,
+          getIncidentResponseUpdatedDateKey,
+        ],
+      });
+
+      const title = getObjectValueByKey<string>(res.data, getIncidentResponseExternalTitleKey);
+      const createdAt = getObjectValueByKey<string>(res.data, getIncidentResponseCreatedDateKey);
+      const updatedAt = getObjectValueByKey<string>(res.data, getIncidentResponseUpdatedDateKey);
+      return { id, title, createdAt, updatedAt };
+    } catch (error) {
+      throw createServiceError(error, `Unable to get incident with id ${id}`);
+    }
+  };
+
+  const createIncident = async ({
+    incident,
+  }: CreateIncidentParams): Promise<ExternalServiceIncidentResponse> => {
+    const { tags, title, description } = incident;
+    try {
+      const res: AxiosResponse = await request({
+        axios: axiosInstance,
+        url: `${createIncidentUrl}`,
+        logger,
+        method: createIncidentMethod,
+        data: renderMustacheStringNoEscape(
```

Comment:
@stephmilovic Forgot to ask, should we `JSON.parse` back to always pass a valid JSON and let Axios do its work based on the headers. We may avoid potential bugs in the future? cc @jonathan-buttner

Reply:
So what you're suggesting is having Axios handle the final stringifying right? I think that's fine.

Reply:
Yes, as we discussed offline if the headers contain `['content-type']: 'application/json'` then we should validate the data by parsing back to JSON and pass to axios a JS object.